### PR TITLE
feat(Message): add thread property

### LINKED ISSUE #290: Support thread in message object
### Summary

Add thread attribute for messsages.

### What is the feature request for?

The core library

### The Problem

In official API we has `thread` attribute, which is not reflected in framework.
https://discord.com/developers/docs/resources/channel#message-object-message-structure

### The Ideal Solution

```py
disnake.Message.thread
```

### The Current Solution

Manual api request to get message and thread data by message id.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

closes #290

Implements disnake.Message.thread which is the thread created on the message.
## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint` or `pre-commit run --all-files`
- [x] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/message.py [hunk 1]
```diff
@@ -1093,6 +1105,11 @@ def _handle_mention_roles(self, role_mentions: List[int]) -> None:
     def _handle_components(self, components: List[ComponentPayload]):
         self.components = [_component_factory(d) for d in components]
 
+    def _handle_thread(self, thread: ThreadPayload) -> None:
+        print("do you even bro")
```
Comment (Victorsitou) [inline disnake/message.py:1109]:
```suggestion
```

Comment (onerandomusername) [reply disnake/message.py:1109]: TLDR no, this method isn't even needed.

File: disnake/message.py [hunk 2]
```diff
@@ -800,6 +800,10 @@ class Message(Hashable):
         .. versionadded:: 2.0
     guild: Optional[:class:`Guild`]
         The guild that the message belongs to, if applicable.
+    thread: Optional[:class:`Thread`]
+        The thread that was started from this message, if applicable.
+
+        ..versionadded: 2.4
```
Comment (Victorsitou) [inline disnake/message.py:806]:
```suggestion
        ..versionadded:: 2.4
```

**Review by Victorsitou — COMMENTED** [review]
Also, why was the `thread` type in `types.Message` deleted?

**Comment by onerandomusername** [comment]
> Also, why was the `thread` type in `types.Message` deleted?

It wasn't?

**Comment by Victorsitou** [comment]
> > Also, why was the `thread` type in `types.Message` deleted?
> 
> It wasn't?

Oh, my bad then, it should be added as well.

File: disnake/message.py [hunk 3]
```diff
@@ -800,6 +800,10 @@ class Message(Hashable):
         .. versionadded:: 2.0
     guild: Optional[:class:`Guild`]
         The guild that the message belongs to, if applicable.
+    thread: Optional[:class:`Thread`]
+        The thread that was started from this message, if applicable.
+
+        ..versionadded:: 2.4
```
Comment (Victorsitou) [inline disnake/message.py:806]: This isn't needed anymore, as it now uses the documentation of the property.

File: disnake/message.py [hunk 4]
```diff
@@ -84,7 +84,7 @@
         MessageReference as MessageReferencePayload,
         Reaction as ReactionPayload,
     )
-    from .types.threads import ThreadArchiveDurationLiteral
+    from .types.threads import Thread as ThreadPayload, ThreadArchiveDurationLiteral
```
Comment (Victorsitou) [inline disnake/message.py:87]: Unused import?

File: disnake/message.py (hunk 3 above)
Comment (onerandomusername) [reply disnake/message.py:806]: Some day I will learn how sphinx works

File: disnake/message.py [hunk 5]
```diff
@@ -1199,6 +1203,15 @@ def jump_url(self) -> str:
         guild_id = getattr(self.guild, "id", "@me")
         return f"https://discord.com/channels/{guild_id}/{self.channel.id}/{self.id}"
 
+    @property
+    def thread(self) -> Optional[Thread]:
+        """
+        Optional[:class:`Thread`]: The thread started from this message. None if no thread was started.
+
+        ..versionadded:: 2.4
```
Comment (Victorsitou) [inline disnake/message.py:1211]:
```suggestion
        .. versionadded:: 2.4
```
sorry :(

**Review by Victorsitou — COMMENTED** [review]
Other than that, looks good to me.

File: disnake/message.py [hunk 6]
```diff
@@ -1199,6 +1203,15 @@ def jump_url(self) -> str:
...
+        """
```
Comment (Victorsitou) [inline disnake/message.py:1212]:
```suggestion
        """
        Optional[:class:`Thread`]: The thread started from this message. ``None`` if no thread was started.

        .. versionadded:: 2.4
        """
```

**Review by Victorsitou — APPROVED** [review]
Looks good to me, the other conversations need to be resolved, by the way.

File: disnake/message.py [hunk 7]
```diff
@@ -896,6 +896,10 @@ def __init__(
         except AttributeError:
             self.guild = state._get_guild(utils._get_as_snowflake(data, "guild_id"))
 
+        if thread_data := data.get("thread"):
```
Comment (shiftinv) [inline disnake/message.py:899]: This should be added to `.types.message._MessageOptional` as well

File: disnake/message.py [hunk 8]
```diff
@@ -1199,6 +1203,15 @@ def jump_url(self) -> str:
         guild_id = getattr(self.guild, "id", "@me")
         return f"https://discord.com/channels/{guild_id}/{self.channel.id}/{self.id}"
 
+    @property
```
Comment (lightmanLP) [inline disnake/message.py:1206]: shouldn't it be a cached property?

**Review by lightmanLP — CHANGES_REQUESTED** [review]

File: disnake/message.py (hunk 8 above)
Comment (onerandomusername) [reply disnake/message.py:1206]: It's possible to delete a thread and create a new thread in which case the cached thread would point to the wrong object.

File: disnake/message.py (hunk 7 above)
Comment (onerandomusername) [reply disnake/message.py:899]: https://github.com/DisnakeDev/disnake/pull/291/commits/b5da05bf5c9ce9460c8a28096400a3ebe1cb4402

**Review by shiftinv — APPROVED** [review]

**Review by EQUENOS — APPROVED** [review]

**Pull request merged**